### PR TITLE
Multi-role, debounced, robust suppress and revive

### PR DESCRIPTION
**Pull request by timcharper**
We introduce a debounced suppress-and-revive mechanism which properly supports multiple Mesos roles (via the UpdateFramework calls).

Further, we introduce the factory pattern to decouple stream wiring logic from component initialization and respective dependencies.

Also, a generated-from-code mechanism is added to generate the visualization docs/scheduler-stream.svg; since this is generated from code, it will not be wrong. The graph is updated when the unit tests are run.

### CODE REVIEW

**Review by kaiwalyajoshi — DISMISSED** [review]
An ASCII graph of the new streams would really help.

**Review by kaiwalyajoshi — APPROVED** [review]

File: core/src/main/scala/com/mesosphere/usi/core/SchedulerFactory.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import com.mesosphere.usi.core.models.commands.SchedulerCommand
+import com.mesosphere.usi.core.revive.SuppressReviveHandler
+import com.mesosphere.usi.core.util.DurationConverters
+import com.mesosphere.usi.metrics.Metrics
+import com.mesosphere.usi.repository.PodRecordRepository
+import org.apache.mesos.v1.scheduler.Protos
+import org.apache.mesos.v1.scheduler.Protos.{Event => MesosEvent}
+import scala.concurrent.Future
+
+trait SchedulerLogicFactory {
```
Comment (jeschkies) [inline core/src/main/scala/com/mesosphere/usi/core/SchedulerFactory.scala:18]: Could you add docs?

File: core/src/main/scala/com/mesosphere/usi/core/SchedulerFactory.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+trait PersistenceFlowFactory {
+  private[usi] def newPersistenceFlow(): Flow[SchedulerEvents, SchedulerEvents, NotUsed]
+}
+
+trait SuppressReviveFactory {
+  def newSuppressReviveFlow: Flow[PodSpecUpdatedEvent, Protos.Call, NotUsed]
+}
+
+case class SchedulerFactory(
```
Comment (jeschkies) [inline core/src/main/scala/com/mesosphere/usi/core/SchedulerFactory.scala:31]: In which case would I not use this factory?

**Review by jeschkies — DISMISSED** [review]
Thanks for the effort. I probably missed some discussion but this pull requests seems a little too big to me. Especially since it introduces multiple new concepts. The factories are orthogonal to the suppress-revive debouncing, isn't it?

File: core/src/main/scala/com/mesosphere/usi/core/revive/SuppressReviveHandler.scala [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  private[this] val reviveCountMetric: Counter = metrics.counter("usi.mesos.calls.revive")
+  private[this] val suppressCountMetric: Counter = metrics.counter("usi.mesos.calls.suppress")
+
+  private def defaultRoles = initialFrameworkInfo.getRolesList.asScala
+
+  /**
+    * Given a stream of PodSpecUpdatedEvent (which gives us the signal of which pods want offers), we emit a snapshot
+    * describing all roles to which the framework is subscribed, and the set of podIds which want offers for each role.
+    */
+  private[revive] def reviveStateFromPodSpecs: Flow[PodSpecUpdatedEvent, PodIdsWantingRoles, NotUsed] =
```
Comment (jeschkies) [inline core/src/main/scala/com/mesosphere/usi/core/revive/SuppressReviveHandler.scala:38]: I have to admit that I'm not a big fan to split this logic out into a flow. The original `suppressAndRevive` was a simple easy to test method. What is the benefit of mixing the logic with streams?

File: core/src/main/scala/com/mesosphere/usi/core/util/RateLimiterFlow.scala [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import java.time.{Clock, Duration => JavaDuration}
+
+import akka.NotUsed
+import akka.stream.scaladsl.Flow
+import akka.stream.stage._
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+
+import scala.concurrent.duration.FiniteDuration
+
+private[usi] class RateLimiterFlow[U] private (rate: FiniteDuration, clock: Clock) extends GraphStage[FlowShape[U, U]] {
```
Comment (jeschkies) [inline core/src/main/scala/com/mesosphere/usi/core/util/RateLimiterFlow.scala:12]: Nice. I actually wanted to point out that this was in Marathon already. What do you think about opening up a PR in Marathon to replace the Marathon implementation with this? We already have a test dependency on USI.

File: core/src/main/scala/com/mesosphere/usi/core/revive/SuppressReviveHandler.scala (hunk 3 above)
Comment (timcharper) [reply core/src/main/scala/com/mesosphere/usi/core/revive/SuppressReviveHandler.scala:38]: Debouncing

File: core/src/main/scala/com/mesosphere/usi/core/util/RateLimiterFlow.scala (hunk 4 above)
Comment (timcharper) [reply core/src/main/scala/com/mesosphere/usi/core/util/RateLimiterFlow.scala:12]: I'd rather not right now

File: core/src/main/scala/com/mesosphere/usi/core/SchedulerFactory.scala (hunk 1 above)
Comment (timcharper) [reply core/src/main/scala/com/mesosphere/usi/core/SchedulerFactory.scala]: You rejected for this?

**Comment by jeschkies** [comment]
> no actual rejection reason

🙁 man. There are a couple:
* The diff is too big and conflates multiple orthogonal changes.
* The interface change for the scheduler creation creates unnecessary work for plugin which has the highest priority as far as I am aware.
* Important parts were not documented.
* I am not a fan of the decision to move important logic into another flow.

**Pull request merged**